### PR TITLE
Lazily import many modules to improve import speed

### PR DESCRIPTION
**Pull request by ashb**
Goal here is to make the healthchecks for triggerer/scheduler quicker by having to do less

Roughly:

- Don't import jsonschema (module _or_ schemas) until we need it - it was surprisingly slow.
- Dont import all models when importing any one -- the only times we need that are when auto-generating migrations, or when checking DB against schema. (That said, accesing any class will cause the SQLA event to import all models, otherwise we often end up with issues due to the relationships on classes)
- Move a few more things to be lazy/delayed import too.

### CODE REVIEW

**Comment by potiuk** [comment]
I really hope one day https://peps.python.org/pep-0690/ will be in place so that we don't have to do all those any more.

File: airflow/jobs/__init__.py [hunk 1]
```diff
@@ -16,8 +16,3 @@
 # specific language governing permissions and limitations
 # under the License.
 #
-import airflow.jobs.backfill_job
```
Comment (mik-laj) [inline airflow/jobs/__init__.py:19]: It is a bit risky. See: https://github.com/apache/airflow/pull/19185

Comment (ashb) [reply airflow/jobs/__init__.py:19]: @mik-laj Yeah I had a feeling something like that might happen - hence why this is in draft and I've got the "load_all_models" fn in airflow.models

**Comment by ashb** [comment]
> I really hope one day https://peps.python.org/pep-0690/ will be in place so that we don't have to do all those any more.

Yeah, that would be nice.

**Comment by uranusjr** [comment]
I recently listened to some of the folks behind PEP 690 discuss lazy imports and came away with the impression that even after it lands Airflow is highly likely not be able to use it. Lazy imports has a lot of implications and PEP 690 quite heavy-handedly make lazy imports the default, and Airflow has so many dependencies I can easily it might _never_ be able to be fully compatible.

**Comment by potiuk** [comment]
> I recently listened to some of the folks behind PEP 690 discuss lazy imports and came away with the impression that even after it lands Airflow is highly likely not be able to use it. Lazy imports has a lot of implications and PEP 690 quite heavy-handedly make lazy imports the default, and Airflow has so many dependencies I can easily it might _never_ be able to be fully compatible.

Yeah. We've probably listened to the same Talk Python episode :).  I think it's not even likely to have PEP 690 around in 3.10 or maybe 3.11. Unless we roll our own sleeves up and help. I think it would be great to actually HELP in PEP 690 by making Airlfow works with lazy imports (maybe by improving PEP 690 implementation as well).

**Comment by ashb** [comment]
Shall I resume this and get the SQLA model loading working do you think?

**Comment by potiuk** [comment]
I think it might be worth it. Would be nice to see how much we gain though in those casese

**Comment by kaxil** [comment]
> Shall I resume this and get the SQLA model loading working do you think?

Yup, lets try it out

**Comment by ashb** [comment]
Wow mypy is confusing/wrong sometimes:

```

airflow/example_dags/example_latest_only_with_trigger.py:39: error: Unexpected
keyword argument "task_id" for "LatestOnlyOperator"  [call-arg]
        latest_only = LatestOnlyOperator(task_id='latest_only')
                      ^
airflow/utils/log/logging_mixin.py: note: "LatestOnlyOperator" defined here
airflow/example_dags/example_latest_only.py:34: error: Unexpected keyword
argument "task_id" for "LatestOnlyOperator"  [call-arg]
        latest_only = LatestOnlyOperator(task_id='latest_only')
                      ^
airflow/utils/log/logging_mixin.py: note: "LatestOnlyOperator" defined here
```

File: airflow/serialization/json_schema.py [hunk 2]
```diff
@@ -19,14 +19,15 @@
...
-
-import jsonschema
+from typing import TYPE_CHECKING, Iterable
 
 from airflow.exceptions import AirflowException
 from airflow.settings import json
 from airflow.typing_compat import Protocol
 
+if TYPE_CHECKING:
+    import jsonschema
```
Comment (uranusjr) [inline airflow/serialization/json_schema.py:29]: Does this change much?

Comment (ashb) [reply airflow/serialization/json_schema.py:29]: It was high up in the list of slow imports!

Comment (uranusjr) [reply airflow/serialization/json_schema.py:29]: A schema validation library being slow… I guess that’s how third-party libraries work.

**Comment by ashb** [comment]
I've found an easy way to make sure all models are loaded when needed:

```
    event.listen(sqlalchemy.orm.mapper, "before_configured", import_all_models, once=True)
```

This will ensure that on the first time any model class is queried or instantiated all other models will be loaded, which ensures we don't get the "class not found in registry" error message, but still means imports are "quick".

**Comment by ashb** [comment]
I wonder why this change causes these mypy errors to appear:

```
tests/system/providers/dbt/cloud/example_dbt_cloud.py:80: error: Unsupported
operand types for >> ("List[object]" and "EmptyOperator")  [operator]
        [get_run_results_artifact, job_run_sensor] >> end
        ^
tests/system/providers/microsoft/azure/example_adf_run_pipeline.py:76: error:
Unsupported operand types for >> ("List[object]" and "EmptyOperator") 
[operator]
        [run_pipeline1, pipeline_run_sensor] >> end
        ^
Found 2 errors in 2 files (checked 2258 source files)

Run mypy for providers.................................................................Failed
- hook id: run-mypy
- exit code: 1

airflow/providers/atlassian/jira/operators/jira.py:70: error: Extra argument
"conf" from **args for "execute" of "JiraOperator"  [misc]
                        resource = self.get_jira_resource_method.execute(*...
                                   ^
airflow/providers/qubole/operators/qubole.py:244: error: Cannot determine type
of "on_failure_callback"  [has-type]
            if self.on_failure_callback is None:
               ^
airflow/providers/qubole/operators/qubole.py:247: error: Cannot determine type
of "on_retry_callback"  [has-type]
            if self.on_retry_callback is None:
```

File: airflow/jobs/__init__.py (hunk 1 above)
Comment (ashb) [reply airflow/jobs/__init__.py:19]: Found a good solution to this that I think is foolproof https://github.com/apache/airflow/pull/24486/commits/685498fa0800c3ca0e51ac41cbd3e863bdab4da6

**Comment by ashb** [comment]
I was testing this with `python -X importtime -c 'import airflow.jobs.scheduler_job' 2>import-before-schedjob.log` and [tuna](https://github.com/nschloe/tuna) for display:

Before:

image


After: 
image

**Comment by ashb** [comment]
I tried adding a `airflow/models/__init__.pyi` containing this:

```python
# MyPy doesn't/can't recoginze the PEP 562 style lazy imports, so we have to
# tell it about the imports

from .dag import DAG
from .base import ID_LEN
from .xcom import XCOM_RETURN_KEY
from .base import Base
from .baseoperator import BaseOperator
from .baseoperator import BaseOperatorLink
from .connection import Connection
from .dagbag import DagBag
from .dag import DagModel
from .dagpickle import DagPickle
from .dagrun import DagRun
from .dag import DagTag
from .db_callback_request import DbCallbackRequest
from .errors import ImportError
from .log import Log
from .mappedoperator import MappedOperator
from .operator import Operator
from .param import Param
from .pool import Pool
from .renderedtifields import RenderedTaskInstanceFields
from .skipmixin import SkipMixin
from .slamiss import SlaMiss
from .taskfail import TaskFail
from .taskinstance import TaskInstance
from .taskreschedule import TaskReschedule
from .trigger import Trigger
from .variable import Variable
from .xcom import XCom
from .taskinstance import clear_task_instances

def import_all_models(): ...
```

But I ended up with loads of errors along the lines of:

```
airflow/www/views.py:109: error: Module "airflow.models" has no attribute "Connection"  [attr-defined]
```

I wonder what is going on ... Any ideas?

File: airflow/utils/cli.py [hunk 3]
```diff
@@ -156,17 +153,6 @@ def _build_metrics(func_name, namespace):
...
-    log = Log(
-        event=f'cli_{func_name}',
-        task_instance=None,
-        owner=metrics['user'],
-        extra=extra,
-        task_id=metrics.get('task_id'),
-        dag_id=metrics.get('dag_id'),
-        execution_date=metrics.get('execution_date'),
-    )
-    metrics['log'] = log
```
Comment (uranusjr) [inline airflow/utils/cli.py:169]: So this metric is lost forever…? (although I don’t really understand how this works)

Comment (ashb) [reply airflow/utils/cli.py:169]: It was only used to write the Log object elsewhere. So I just created it there instead.

("metric" was an odd name for this variable in the first place)

File: airflow/models/__init__.py [hunk 4]
```diff
@@ -78,3 +50,94 @@
...
+    'TaskReschedule': 'airflow.models.taskreschedule',
+    'Trigger': 'airflow.models.trigger',
+    'Variable': 'airflow.models.variable',
+    'XCom': 'airflow.models.xcom',
+    'clear_task_instances': 'airflow.models.taskinstance',
+}
+
+STATICA_HACK = True
+globals()['kcah_acitats'[::-1].upper()] = False
+if STATICA_HACK:  # pragma: no cover
```
Comment (uranusjr) [inline airflow/models/__init__.py]: Would `TYPE_CHECKING` not work here?

Comment (ashb) [reply airflow/models/__init__.py]: oh yes, probably! Testing

Comment (ashb) [reply airflow/models/__init__.py]: Yup this works.

**Review by jedcunningham — APPROVED** [review]

**Review by potiuk — COMMENTED** [review]
Nice one :)

**Pull request merged**